spec/integration/app_spec.rb

```diff
@@ -0,0 +1,221 @@
+      repo.add_user(user)
+      response = post(
+        '/login',
+        email: 'Jmarie2002@example.com',
+        password: 'foo')
+      get('/logout')
+      expect(last_request.env["rack.session"]).to be_empty
+    end
+  end
+
+  context "POST /login" do
+    it "returns a success page" do
+      repo = UserRepository.new
+      user = User.new
+      user.id = 4
+      user.name = 'Janie Marie'
+      user.username = 'JMarie1'
+      user.password = 'foo'
+      user.email = 'Jmarie2002@example.com'
+      repo.add_user(user)
+      response = post(
+        '/login',
+        email: 'Jmarie2002@example.com',
+        password: 'foo')
+      expect(response.status).to eq 200
+      expect(response.body).to include('<h1>Login successfull</h1>')
+      expect(response.body).to include('<a href="/logged_in_home">Continue to Home</a>')
+    end
+
+    it "returns failure page if password is incorrect" do
+      repo = UserRepository.new
+      user = User.new
+      user.id = 4
+      user.name = 'Janie Marie'
+      user.username = 'JMarie1'
+      user.password = 'foo'
+      user.email = 'Jmarie2002@example.com'
+      repo.add_user(user)
+      response = post(
+        '/login',
+        email: 'Jmarie2002@example.com',
+        password: 'bar')
+      expect(response.status).to eq 200
+      expect(response.body).to include('<h1>Username or password is incorrect</h1>')
+      expect(response.body).to include('<a href="/login">Try again </a>')
+    end
+
+    it "Responds with 400 if input is empty string" do
+      response = post('/post', { email: '' }, { 'rack.session' => { user_id: 1 } })
+      expect(response.status).to eq(400)
+    end
+
+    it "Responds with 400 if input is nil" do
+      response = post('/post', { email: nil }, { 'rack.session' => { user_id: 1 } })
+      expect(response.status).to eq(400)
+    end
+  end
+
+  context "POST /post" do
+    it 'returns 200 OK' do
+      # Assuming the post with id 1 exists.
+      response = post('/post', { content: 'A good day :)' }, { 'rack.session' => { user_id: 1 } })
+      expect(response.status).to eq(200)
+    end
+
+    it 'database includes new post' do
+      response = post('/post', { content: 'A good day :)' }, { 'rack.session' => { user_id: 1 } })
+      get_response = get('/post/6')
+      expect(get_response.status).to eq(200)
+      expect(get_response.body).to include "<h1>Just finished a great workout</h1>"
+      expect(get_response.body).to include "Time posted: 2022-09-06 07:30:00"
+      expect(get_response.body).to include "User: bobj"
+    end
+
+    it 'returns success message' do
+      response = post('/post', { content: 'A good day :)' }, { 'rack.session' => { user_id: 1 } })
+      expect(response.status).to eq(200)
+      expect(response.body).to include("<h1>Peep posted :)</h1>")
+      expect(response.body).to include('<a href="/logged_in_home">Return to home </a>')
+    end
+
+    it "Responds with 400 if input is empty string" do
```

Comment:
testing for response status is very good and shows that you have in depth tests for a wide variety of things and scenarios